### ISSUE TITLE
NS5x_NS7xPU ADL: firmware defaults to maximum CD clock, preventing flicker-free booting

### ISSUE THREAD

**Issue by XGIVolari** [Dasharo/dasharo-issues#493]
**Dasharo version**

Notebook NS5x_NS7xPU/NS5x_NS7xPU, BIOS Dasharo (coreboot+UEFI) v1.6.0 03/30/2023

**Affected component(s) or functionality**

GPU

**Brief summary**

If the hardware is initialized correctly, the i915 kernel driver can skip all boot-time modesetting operations, allowing the boot process to complete 100% flicker free. On ADL-P platforms, this requires the core display clock to be set to the right value (in this case 179200 kHz). However, as of release v1.6.0, the firmware initializes the CDCLK at the maximum supported value of 652800 kHz, forcing the kernel to perform modesetting anyways. 

**How reproducible**

always

**How to reproduce**

1. boot with the parameter "drm.debug=0x1e"
2. look for the following lines:
```
[    2.393620] i915 0000:00:02.0: [drm:intel_cdclk_init_hw [i915]] Current CDCLK 652800 kHz, VCO 1305600 kHz, ref 38400 kHz, bypass 19200 kHz, voltage level 3
[    2.397149] i915 0000:00:02.0: [drm:intel_cdclk_dump_config [i915]] Current CDCLK 652800 kHz, VCO 1305600 kHz, ref 38400 kHz, bypass 19200 kHz, voltage level 3
[    2.397267] i915 0000:00:02.0: [drm:intel_update_max_cdclk [i915]] Max CD clock rate: 652800 kHz
[    2.450273] i915 0000:00:02.0: [drm:intel_modeset_calc_cdclk [i915]] Modeset required for cdclk change
[    3.232233] i915 0000:00:02.0: [drm:intel_set_cdclk [i915]] Changing CDCLK to 179200 kHz, VCO 537600 kHz, ref 38400 kHz, bypass 19200 kHz, voltage level 0
```

**Expected behavior**

The firmware defaults do a CDCLK of 179200 kHz, allowing the i915 driver to skip a modesetting operation. The system boots flicker-free. (with the required kernel patches, see below)

**Actual behavior**

The firmware sets the CDCLK to the maximum value of 652800 kHz, forcing the i915 driver to perform a full modeset in order to change it. The screen flickers at boot.

**Additional context**

Tested on Gentoo using the precompiled kernel version 6.5.3.

On the kernel side, there's an additonal patch series that's necessary for flicker-free booting to work. This is however unrelated to the firmware.
https://lists.freedesktop.org/archives/intel-gfx-trybot/2023-August/135601.html

**Solutions you have tried**
working around the issue by disabling various i915 power saving features (unsuccessfully)

**Comment by XGIVolari** [comment]
Unfortunately, I don't know for sure. However, the [i915 bug tracker](https://gitlab.freedesktop.org/drm/intel/-/issues) has plenty of drm debug logs captured from ADL-P platforms that show that other devices initialize the CDCLK at 179200 kHz as well. For example:
[#8312](https://gitlab.freedesktop.org/drm/intel/-/issues/8312): [external-thinkvision-x11-bug.log](https://gitlab.freedesktop.org/drm/intel/uploads/fb6e14219bb38e67703986f8fcc62f9e/external-thinkvision-x11-bug.log)
[#8097](https://gitlab.freedesktop.org/drm/intel/-/issues/8097): [dmesg.txt](https://gitlab.freedesktop.org/drm/intel/uploads/1899be8b909061d467569bcc94effa19/dmesg.txt)
[#7687](https://gitlab.freedesktop.org/drm/intel/-/issues/7687): [drm-tip-full-log.txt](https://gitlab.freedesktop.org/drm/intel/uploads/c6f79b0a8380e76edbd488c90df64844/drm-tip-full-log.txt)

In fact, in my (by no means comprehensive) search for more info about this, I haven't found a single ADL-P log yet that lists anything other than 179200 kHz as the default CDCLK.

**Comment by miczyg1** [comment]
@XGIVolari do you have any insight why Linux does not use max CDclock by default? At least this is what Intel uses as default in their reference BIOS.

**Comment by XGIVolari** [comment]
This seems to be a power saving feature. At least the [documentation](https://github.com/torvalds/linux/blob/2cf0f715623872823a72e451243bbf555d10d032/drivers/gpu/drm/i915/display/intel_cdclk.c#L44) inside the i915 source code suggest as much:
```
 * CDCLK clocks most of the display pipe logic, and thus its frequency
 * must be high enough to support the rate at which pixels are flowing
 * through the pipes. Downscaling must also be accounted as that increases
 * the effective pixel rate.
 *
 * On several platforms the CDCLK frequency can be changed dynamically
 * to minimize power consumption for a given display configuration.
 * Typically changes to the CDCLK frequency require all the display pipes
 * to be shut down while the frequency is being changed.
 ```

**Comment by miczyg1** [comment]
@XGIVolari thanks for that. For me it is a standard practice to set the highest performance available by default for the silicon in the firmware. It is kind of unexpected to set the lowest possible CDclock immediately. If anything, I would expect Linux to set it when system is idling or such. If I understand correctly, lower CDclock means lower bandwidth of the display pipes, and as a result, lower FPS on LCD matrix.

**Comment by XGIVolari** [comment]
Glad I can help. Does bandwith requirement increase if the amount of FPS the GPU renders is higher than the display's refresh rate? AFAIU, any bandwith higher than what is necessary to drive all connected displays at maximum refresh rate is just wasted power consumption. And since ADL-P can drive up to 4 4K HDR displays simultaneously, such a low CDCLK still provides plenty of bandwidth to drive a single standard 1080p screen without any performance downsides. If more bandwith is required (e.g. by connecting an additional display), the kernel raises the CD clock at the "expense" of potentially flickering briefly.
```
	 * Ie. we only ever increase the cdclk due to bandwidth
	 * requirements. This can reduce back and forth
	 * display blinking due to constant cdclk changes.
```

**Comment by mkopec** [comment]
>  Does bandwith requirement increase if the amount of FPS the GPU renders is higher than the display's refresh rate?

No, the pixel clock needs to be just enough for the selected mode, 1080p60 for the internal display in this case. Should not affect rendering performance at all.

**Comment by mkopec** [comment]
@miczyg1 looks like VBT tells GOP whetner to downclock CD Clock at runtime, see https://github.com/intel/FSP/blob/master/AlderLakeFspBinPkg/Client/AlderLakeP/SampleCode/Vbt/Vbt.json line 23542

**Comment by pietrushnic** [comment]
@XGIVolari thanks for reporting the issues. @mkopec does it make sense for @XGIVolari to test your changes?

**Comment by mkopec** [comment]
@pietrushnic Sure, @XGIVolari if you are feeling like testing bleeding edge firmware, you can get a preview version from https://github.com/Dasharo/coreboot/actions/runs/6252453730?pr=398 and follow the instructions at https://docs.dasharo.com/dasharo-tools-suite/documentation/#local-firmware-update to install and test. So far I've only verified that Windows boots correctly.

**Comment by mkopec** [comment]
Perfect, thanks for testing! We'll merge the fix and it'll be included in an upcoming official firmware update.
